icu4c/source/common/uloc.cpp

```diff
@@ -1324,13 +1324,31 @@ _getVariant(const char* localeID,
             char prev,
             ByteSink* sink,
             const char** pEnd,
-            bool needSeparator) {
+            bool needSeparator,
+            UErrorCode& status) {
+    if (U_FAILURE(status)) return;
+    const char* start = localeID;
     bool hasVariant = false;
     if (pEnd != nullptr) { *pEnd = localeID; }
 
+    // Reasonable upper limit for variants
+    constexpr int32_t MAX_NUMBER_OF_VARIANTS = 20;
+    constexpr int32_t MAX_VARIANT_LENGTH = 8;
+    constexpr int32_t MAX_VARIANTS_LENGTH = (MAX_VARIANT_LENGTH + 1) * MAX_NUMBER_OF_VARIANTS;
```

Comment:
This is just an estimation since we do not know the exact legnth of each variant. If each variant value is only length of three, we could fit 45 variants here so I do not think we need it to be exact since neither  MAX_VARIANTS_LENGTH = (MAX_VARIANT_LENGTH + 1) * MAX_NUMBER_OF_VARIANTS 
nor
MAX_VARIANTS_LENGTH = (MAX_VARIANT_LENGTH + 1) * MAX_NUMBER_OF_VARIANTS - 1
could guarantee there are at most 20 variants in this.
and due to the acceptable grammar of variant, if all the variant are lenght of 8 there will be no syntax valid variants string be exact 180 long anyway.